clients/hadoopfs/src/main/java/io/lakefs/FileSystemTracer.java

```diff
@@ -0,0 +1,489 @@
+        IOException s3aException = null;
+        Path s3aPath = translateLakeFSPathToS3APath(f);
+        try {
+            lakefsRes = lfsFileSystem.open(f, bufferSize);
+        } catch (IOException e) {
+            lakeFSException = e;
+            LOG.error("[open] Can't open {} with lakeFSFileSystem, exception {}", f, e.getMessage());
+        }
+        try {
+            s3aRes = s3AFileSystem.open(translateLakeFSPathToS3APath(f), bufferSize);
+        } catch (IOException e) {
+            s3aException = e;
+            LOG.error("[open] Can't open {} with S3AFileSystem, exception {}", s3aPath, e.getMessage());
+        }
+
+        if (useLakeFSFileSystemResults && lakeFSException != null) {
+            LOG.trace("[open] exception by lakeFSFileSystem");
+            throw lakeFSException;
+        }
+        if (!useLakeFSFileSystemResults && s3aException != null) {
+            LOG.trace("[open] exception by S3AFileSystem");
+            throw s3aException;
+        }
+
+        LOG.trace("{}[open] lakefs: {}, s3a: {}", RESULTS_COMPARISON, lakefsRes, s3aRes);
+        if (useLakeFSFileSystemResults) {
+            return lakefsRes;
+        }
+        return s3aRes;
+    }
+
+    @Override
+    public FSDataOutputStream create(Path f, FsPermission permission, boolean overwrite, int bufferSize,
+                                     short replication, long blockSize, Progressable progress) throws IOException {
+        LOG.trace("create(Path {}, permission {}, overwrite {}, bufferSize {}, replication {}, blockSize {}, progress {})",
+                f, permission, overwrite, bufferSize, replication, blockSize, progress);
+        FSDataOutputStream lakeFSStream = null;
+        FSDataOutputStream s3aStream = null;
+        IOException lakeFSException = null;
+        IOException s3aException = null;
+        Path s3aPath = translateLakeFSPathToS3APath(f);
+        try {
+            lakeFSStream = lfsFileSystem.create(f, permission, overwrite, bufferSize, replication, blockSize, progress);
+        } catch (IOException e) {
+            lakeFSException = e;
+            LOG.error("[create] Can't create {} with lakeFSFileSystem, exception {}", f, e.getMessage());
+        }
+        try {
+            s3aStream = s3AFileSystem.create(s3aPath, permission, overwrite, bufferSize,
+                replication, blockSize, progress);
+        } catch (IOException e) {
+            s3aException = e;
+            LOG.error("[create] Can't create {} with S3AFileSystem, exception {}", s3aPath, e.getMessage());
+        }
+
+        if (useLakeFSFileSystemResults && lakeFSException != null) {
+            LOG.trace("[create] exception by lakeFSFileSystem");
+            s3aStream.close();
+            throw lakeFSException;
+        }
+        if (!useLakeFSFileSystemResults && s3aException != null) {
+            LOG.trace("[create] exception by S3AFileSystem");
+            lakeFSStream.close();
+            throw s3aException;
+        }
```

Comment:
my concern here is when we throw and choose to close - it doesn't mean that we managed to open the other stream and we probably get null pointer exception when we try to close it.